### PR TITLE
feat: Sorting for search results

### PR DESCRIPTION
**Pull request by adilansari**
Interface to allow users to specify sort order for search results.

```java
TigrisSort asc = Sort.ascending("a");
TigrisSort desc = Sort.descending("b");

SearchRequest req = SearchRequest.matchAll().withSort(asc, desc).build();
```

OR

```java
TigrisSort asc = Sort.ascending("a");
TigrisSort desc = Sort.descending("b");
SortingOrder order = SortingOrder.withOrder(asc, desc).build()

SearchRequest req = SearchRequest.newBuilder().withQ("running").withSort(order).build();
```

### CODE REVIEW

File: client/src/main/java/com/tigrisdata/db/client/FieldSort.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+package com.tigrisdata.db.client;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/** Represents a collection field and corresponding sort order */
+public final class FieldSort implements TigrisSort {
```
Comment (adilansari) [inline client/src/main/java/com/tigrisdata/db/client/FieldSort.java:23]: In future we can extend this further 

```
class GeoDistanceSort implements TigrisSort {}
Sort.geoDistance(String fieldName, GeoLocation loc)
```

File: client/src/main/java/com/tigrisdata/db/client/TigrisSort.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+/** Represents the sorting order */
+public interface TigrisSort {
+
+  /**
+   * Sort order representation as Key, Value pairs
+   *
+   * @return non-null immutable {@link Map}
+   */
+  public Map<String, Object> toMap();
```
Comment (JigarJoshi) [inline client/src/main/java/com/tigrisdata/db/client/TigrisSort.java:27]: remove `public` interface methods are implicitly public. 

toMap();  is an internal function and shouldn't be visible to user.

File: client/src/main/java/com/tigrisdata/db/client/TigrisSort.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+  /**
+   * Gets collection field name that this order represents
+   *
+   * @return {@link String}
+   */
+  public String getFieldName();
```
Comment (JigarJoshi) [inline client/src/main/java/com/tigrisdata/db/client/TigrisSort.java:34]: remove `public`, should we redefine this interface to contain two methods

getFieldName(); getSortOrder(); ?

File: client/src/main/java/com/tigrisdata/db/client/SortingOperator.java [hunk 4]
```diff
@@ -12,9 +12,19 @@
  * limitations under the License.
  */
 
-package com.tigrisdata.db.client.search;
+package com.tigrisdata.db.client;
 
-import com.tigrisdata.db.client.JSONSerializable;
+enum SortingOperator {
```
Comment (JigarJoshi) [inline client/src/main/java/com/tigrisdata/db/client/SortingOperator.java:17]: nit: `SortingOperator` --> `SortOrder`

File: client/src/main/java/com/tigrisdata/db/client/Sort.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+public final class Sort {
+  private Sort() {}
+
+  /**
+   * Creates ascending sort order for given key
+   *
+   * @param key collection field name
+   * @return constructed {@link FieldSort} of ascending order
+   */
+  public static FieldSort ascending(String key) {
```
Comment (JigarJoshi) [inline client/src/main/java/com/tigrisdata/db/client/Sort.java:27]: nit: `key` --> `fieldName`

File: client/src/main/java/com/tigrisdata/db/client/Sort.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    return new FieldSort(key, SortingOperator.ASC);
+  }
+
+  /**
+   * Creates descending sort order for given key
+   *
+   * @param key collection field name
+   * @return constructed {@link FieldSort} of descending order
+   */
+  public static FieldSort descending(String key) {
```
Comment (JigarJoshi) [inline client/src/main/java/com/tigrisdata/db/client/Sort.java:37]: nit: `key` --> `fieldName`

File: client/src/main/java/com/tigrisdata/db/client/FieldSort.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public String getFieldName() {
+    return fieldName;
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
Comment (JigarJoshi) [inline client/src/main/java/com/tigrisdata/db/client/FieldSort.java:56]: nit: in other places in this library. I have been using IDE generated equals and hashCode (Java 7+) with java.util.Objects

File: client/src/main/java/com/tigrisdata/db/client/search/SearchRequest.java [hunk 8]
```diff
@@ -214,17 +215,29 @@ public Builder withFacetQuery(FacetQuery facetQuery) {
...
-     * Optional - Sets the SortOrder to sorts search results according to specified attributes and
+     * Optional - Sets the SortingOrder to sort search results according to specified attributes and
      * indicated order
      *
-     * @param sortOrder {@link SortOrder}
+     * @param sortOrder {@link SortingOrder}
      * @return {@link SearchRequest.Builder}
      */
-    public Builder withSortOrder(SortOrder sortOrder) {
+    public Builder withSort(SortingOrder sortOrder) {
```
Comment (JigarJoshi) [inline client/src/main/java/com/tigrisdata/db/client/search/SearchRequest.java:224]: What are the use cases when this method is useful vs the following?

Comment (JigarJoshi) [reply client/src/main/java/com/tigrisdata/db/client/search/SearchRequest.java:224]: can you add a rpc method test to exercise the full trip?

Comment (adilansari) [reply client/src/main/java/com/tigrisdata/db/client/search/SearchRequest.java:224]: Allow sharing of `SortingOrder` between search requests. Or if we introduce sorting for read.

Comment (JigarJoshi) [reply client/src/main/java/com/tigrisdata/db/client/search/SearchRequest.java:224]: makes sense.

**Review by JigarJoshi — APPROVED** [review]

**Pull request merged**